scripts/lib/CIME/XML/generic_xml.py

```diff
@@ -35,31 +57,169 @@ def __init__(self, infile=None, schema=None):
             expect("$" not in infile,"File path not fully resolved {}".format(infile))
 
             self.filename = infile
-            root = ET.Element("xml")
-            self.root = ET.SubElement(root, "file")
-            self.root.set("id", os.path.basename(infile))
-            self.root.set("version", "2.0")
+            root = _Element(ET.Element("xml"))
+            if root_name_override:
+                self.root = self.make_child(root_name_override, root=root, attributes=root_attrib_override)
+            else:
+                self.root = self.make_child("file", root=root, attributes={"id":os.path.basename(infile), "version":"2.0"})
+
             self.tree = ET.ElementTree(root)
 
     def read(self, infile, schema=None):
         """
         Read and parse an xml file into the object
         """
-        logger.debug("read: " + infile)
+        logger.debug("read: " + str(infile))
+
         with open(infile, 'r') as fd:
-            if self.tree:
-                self.root.append(ET.parse(fd).getroot())
-            else:
-                self.tree = ET.parse(fd)
-                self.root = self.tree.getroot()
+            self.read_fd(fd)
 
         if schema is not None and self.get_version() > 1.0:
             self.validate_xml_file(infile, schema)
 
         logger.debug("File version is {}".format(str(self.get_version())))
 
+    def read_fd(self, fd):
+        if self.tree:
+            self.add_child(_Element(ET.parse(fd).getroot()))
+        else:
+            self.tree = ET.parse(fd)
+            self.root = _Element(self.tree.getroot())
+
+    #
+    # API for individual node operations
```

Comment:
Important, here is our API for node operations.

We support the concepts:
* Nodes
* Nodes can have a key:value dict of attributes
* Nodes have a name
* Node can have text
* Nodes can have child nodes.